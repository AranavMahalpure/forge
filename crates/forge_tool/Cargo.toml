[package]
name = "forge_tool"
version = "0.1.0"
edition = "2021"

[dependencies]
dissimilar = "1.0"
tokio = { version = "1.0", features = ["full"] }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
anyhow = "1.0"
async-trait = "0.1"
tracing = "0.1"
colorize = "0.1.0"
Inflector = "0.11.4"
forge_tool_macros = { path = "../forge_tool_macros" }
schemars = "0.8.21"
regex = "1.10.2"
walkdir = "2.4.0"
glob = "0.3.1"
tree-sitter = "0.24.0"
tree-sitter-rust = "0.23"
tree-sitter-javascript = "0.23"
tree-sitter-python = "0.23"
forge_walker = { path = "../forge_walker" }
forge_domain = { path = "../forge_domain" }
tempfile = "3.8.1"
streaming-iterator = "0.1.9"
uuid = { version = "1.0", features = ["v4", "serde"] }

[dev-dependencies]
insta = "1.34.0"
pretty_assertions = "1.4.1"
