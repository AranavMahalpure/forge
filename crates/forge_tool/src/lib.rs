mod approve;
mod ask;
mod fetch;
mod fs;
mod learning;
mod outline;
mod patch;
mod select;
mod shell;
mod syn;
mod think;

use std::sync::Arc;

use approve::Approve;
use fetch::Fetch;
use forge_domain::{EmbeddingsRepository, Tool};
use fs::*;
use learning::Learning;
use outline::Outline;
use patch::ApplyPatch;
use select::SelectTool;
use shell::Shell;
use think::Think;

pub fn tools(learning_embedding_idx: Arc<dyn EmbeddingsRepository>) -> Vec<Tool> {
    vec![
        Approve.into(),
        FSRead.into(),
        FSWrite.into(),
        FSList.into(),
        FSSearch.into(),
        FSFileInfo.into(),
        ApplyPatch.into(),
        Outline.into(),
        SelectTool.into(),
        Shell::default().into(),
        Think::default().into(),
        Fetch::default().into(),
        Learning::new(learning_embedding_idx).into(),
    ]
}

#[cfg(test)]
mod tests {
    use super::*;
    use learning::tests::MockEmbeddingsRepository;

    #[test]
    fn test_tool_description_length() {
        const MAX_DESCRIPTION_LENGTH: usize = 1024;

        println!("\nTool description lengths:");

        let mut any_exceeded = false;
        for tool in tools(Arc::new(MockEmbeddingsRepository::default())) {
            let desc_len = tool.definition.description.len();
            println!(
                "{:?}: {} chars {}",
                tool.definition.name,
                desc_len,
                if desc_len > MAX_DESCRIPTION_LENGTH {
                    "(!)"
                } else {
                    ""
                }
            );

            if desc_len > MAX_DESCRIPTION_LENGTH {
                any_exceeded = true;
            }
        }

        assert!(
            !any_exceeded,
            "One or more tools exceed the maximum description length of {}",
            MAX_DESCRIPTION_LENGTH
        );
    }
}
