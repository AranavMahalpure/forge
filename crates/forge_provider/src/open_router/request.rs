use forge_tool::{ToolDefinition, ToolName};
use serde::{Deserialize, Serialize};

use crate::{ModelId, Request, RequestMessage, Role, ToolUseId};

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct TextContent {
    // TODO: could be an enum
    pub r#type: String,
    pub text: String,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct ImageContentPart {
    pub r#type: String,
    pub image_url: ImageUrl,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct ImageUrl {
    pub url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub detail: Option<String>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct OpenRouterMessage {
    pub role: Role,
    pub content: MessageContent,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<ToolName>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<ToolUseId>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(untagged)]
pub enum MessageContent {
    Text(String),
    Parts(Vec<ContentPart>),
}

impl MessageContent {
    pub fn cached(self) -> Self {
        match self {
            MessageContent::Text(text) => MessageContent::Parts(vec![ContentPart::Text {
                text,
                cache_control: Some(CacheControl { type_: CacheControlType::Ephemeral }),
            }]),
            _ => self,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ContentPart {
    Text {
        text: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        cache_control: Option<CacheControl>,
    },
    ImageUrl {
        image_url: ImageUrl,
    },
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct CacheControl {
    #[serde(rename = "type")]
    pub type_: CacheControlType,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "snake_case")]
pub enum CacheControlType {
    Ephemeral,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct FunctionDescription {
    pub description: Option<String>,
    pub name: String,
    pub parameters: serde_json::Value,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct OpenRouterTool {
    // TODO: should be an enum
    pub r#type: String,
    pub function: FunctionDescription,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub enum ToolChoice {
    None,
    Auto,
    Function { name: String },
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct ResponseFormat {
    pub r#type: String,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct Prediction {
    pub r#type: String,
    pub content: String,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct ProviderPreferences {
    // Define fields as necessary
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct OpenRouterRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub messages: Option<Vec<OpenRouterMessage>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt: Option<String>,
    pub model: ModelId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response_format: Option<ResponseFormat>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stop: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tools: Option<Vec<OpenRouterTool>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_choice: Option<ToolChoice>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub seed: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_p: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_k: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frequency_penalty: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub presence_penalty: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub repetition_penalty: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub logit_bias: Option<std::collections::HashMap<u32, f32>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_logprobs: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_p: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_a: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prediction: Option<Prediction>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transforms: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub models: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub route: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider: Option<ProviderPreferences>,
}

impl From<ToolDefinition> for OpenRouterTool {
    fn from(value: ToolDefinition) -> Self {
        OpenRouterTool {
            r#type: "function".to_string(),
            function: FunctionDescription {
                description: Some(value.description),
                name: value.name.into_string(),
                parameters: serde_json::to_value(value.input_schema).unwrap(),
            },
        }
    }
}

impl From<Request> for OpenRouterRequest {
    fn from(request: Request) -> Self {
        OpenRouterRequest {
            messages: {
                let messages = request
                    .messages
                    .into_iter()
                    .map(OpenRouterMessage::from)
                    .collect::<Vec<_>>();

                Some(insert_cache(messages))
            },
            tools: {
                let tools = request
                    .tools
                    .into_iter()
                    .map(OpenRouterTool::from)
                    .collect::<Vec<_>>();
                if tools.is_empty() {
                    None
                } else {
                    Some(tools)
                }
            },
            model: request.model,
            prompt: Default::default(),
            response_format: Default::default(),
            stop: Default::default(),
            stream: Default::default(),
            max_tokens: Default::default(),
            temperature: Default::default(),
            tool_choice: Default::default(),
            seed: Default::default(),
            top_p: Default::default(),
            top_k: Default::default(),
            frequency_penalty: Default::default(),
            presence_penalty: Default::default(),
            repetition_penalty: Default::default(),
            logit_bias: Default::default(),
            top_logprobs: Default::default(),
            min_p: Default::default(),
            top_a: Default::default(),
            prediction: Default::default(),
            transforms: Default::default(),
            models: Default::default(),
            route: Default::default(),
            provider: Default::default(),
        }
    }
}

impl From<RequestMessage> for OpenRouterMessage {
    fn from(value: RequestMessage) -> Self {
        match value {
            RequestMessage::Chat(chat_message) => OpenRouterMessage {
                role: chat_message.role.into(),
                content: MessageContent::Text(chat_message.content),
                name: None,
                tool_call_id: None,
            },
            RequestMessage::ToolResult(tool_result) => OpenRouterMessage {
                role: Role::Tool,
                content: MessageContent::Text(serde_json::to_string(&tool_result.content).unwrap()),
                name: Some(tool_result.tool_name),
                tool_call_id: tool_result.tool_use_id,
            },
        }
    }
}

/// Inserts cache control information into system messages
/// NOTE: We need to add more caching as the context grows larger
fn insert_cache(mut message: Vec<OpenRouterMessage>) -> Vec<OpenRouterMessage> {
    for message in message.iter_mut() {
        if message.role == Role::System {
            message.content = message.content.clone().cached();
        }
    }

    message
}
