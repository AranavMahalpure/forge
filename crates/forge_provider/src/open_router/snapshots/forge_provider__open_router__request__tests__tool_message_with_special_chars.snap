---
source: crates/forge_provider/src/open_router/request.rs
expression: router_message
snapshot_kind: text
---
{
  "role": "tool",
  "content": "<tool_result><tool_name>html_tool</tool_name><success><attributes><data-test><test>&value</test></data-test><style>color: blue; font-size: 12px;</style></attributes><elements><span></elements><elements><br/></elements><elements><hr></elements><html><div class=\"container\"><p>Hello <World></p></div></html></success></tool_result>",
  "name": "html_tool",
  "tool_call_id": "456"
}
